src/java/org/apache/cassandra/auth/MutualTlsAuthenticator.java

```diff
@@ -0,0 +1,207 @@
+                                                              Arrays.asList("", AuthConfig.class.getPackage().getName()));
+        checkMtlsConfigurationIsValid(DatabaseDescriptor.getRawConfig());
+        AuthCacheService.instance.register(identityCache);
+    }
+
+    @Override
+    public boolean requireAuthentication()
+    {
+        return true;
+    }
+
+    @Override
+    public Set<? extends IResource> protectedResources()
+    {
+        return ImmutableSet.of(DataResource.table(SchemaConstants.AUTH_KEYSPACE_NAME, AuthKeyspace.ROLES));
+    }
+
+    @Override
+    public void validateConfiguration() throws ConfigurationException
+    {
+
+    }
+
+    @Override
+    public void setup()
+    {
+        identityCache.warm();
+    }
+
+    @Override
+    public SaslNegotiator newSaslNegotiator(InetAddress clientAddress)
+    {
+        return null;
+    }
+
+    @Override
+    public SaslNegotiator newSaslNegotiator(InetAddress clientAddress, Certificate[] certificates)
+    {
+        return new CertificateNegotiator(certificates);
+    }
+
+    @Override
+    public AuthenticatedUser legacyAuthenticate(Map<String, String> credentials) throws AuthenticationException
+    {
+        return null;
+    }
+
+    @VisibleForTesting
+    class CertificateNegotiator implements SaslNegotiator
+    {
+        private final Certificate[] clientCertificateChain;
+
+        private CertificateNegotiator(final Certificate[] clientCertificateChain)
+        {
+            this.clientCertificateChain = clientCertificateChain;
+        }
+
+        @Override
+        public byte[] evaluateResponse(byte[] clientResponse) throws AuthenticationException
+        {
+            return null;
+        }
+
+        @Override
+        public boolean isComplete()
+        {
+            return true;
+        }
+
+        @Override
+        public AuthenticatedUser getAuthenticatedUser() throws AuthenticationException
+        {
+            if (!certificateValidator.isValidCertificate(clientCertificateChain))
+            {
+                String message = "Invalid or not supported certificate";
+                nospamLogger.error(message);
+                throw new AuthenticationException(message);
+            }
+
+            final String identity = certificateValidator.identity(clientCertificateChain);
+            if (StringUtils.isEmpty(identity))
+            {
+                String msg = "Unable to extract client identity from certificate for authentication";
+                nospamLogger.error(msg);
+                throw new AuthenticationException(msg);
+            }
+            String role = identityCache.get(identity);
+            if (role == null)
+            {
+                String msg = "None of the users " + identity + " are authorized";
+                nospamLogger.error(msg);
+                throw new AuthenticationException(msg);
```

Comment:
nit: this is a possibility for log injection - the identity provided in the client certificate could include text to confuse a log scanner,  e.g. an identity of "%property{my_sensitive_property}" could cause it to be logged and be exploited if any more log4shell type vulnerabilities are discovered.

You could use `org.slf4j.helpers.MessageFormatter` to create the `msg` instead, that would also reduce the size of the map nospamLogger might create for previous timestamps.
```suggestion
                String msg = "None of the users {} are authorized";
                nospamLogger.error(msg, identity);
                throw new AuthenticationException(MessageFormatter.format(msg, identity).getMessage());
```

Reply:
`Certificate identity '{}' not authorized` might also make a clearer message in logs and exceptions - as I think there's only one identity.